src/main/scala/uk/gov/hmrc/cache/repository/CacheRepository.scala

```diff
@@ -42,44 +45,52 @@ object CacheRepository extends MongoDbConnection {
 class CacheMongoRepository(collName: String, override val expireAfterSeconds: Long, cacheFormats: Format[Cache] = Cache.mongoFormats)(implicit mongo: () => DB, ec: ExecutionContext)
   extends ReactiveRepository[Cache, Id](collName, mongo, cacheFormats, Id.idFormats)
     with CacheRepository with TTLIndexing[Cache]
-    with AtomicUpdate[Cache]
     with BSONBuilderHelpers {
 
 
-  final val AtomicId="atomicId"
-  final val Id="_id"
+  final val AtomicId = "atomicId"
+  final val Id = "_id"
 
-  private def findByIdBSON(id: Id) = BSONDocument(Id -> BSONString(id.id))
-
-  private def modifierForCrudCredentialsBSON(time: Long): BSONDocument = BSONDocument(
-    "$set" -> BSONDocument("modifiedDetails.lastUpdated" -> BSONDateTime(time)),
-    "$setOnInsert" -> BSONDocument("modifiedDetails.createdAt" -> BSONDateTime(time))
-  )
+  import ReactiveMongoFormats._
 
   override def createOrUpdate(id: Id, key: String, toCache: JsValue): Future[DatabaseUpdate[Cache]] = {
 
     withCurrentTime {
 
       implicit time =>
 
-        val modifiers = List(
-          // Set the document and overwrite if already exists.
-          set(BSONDocument(s"${Cache.DATA_ATTRIBUTE_NAME}.$key" -> BSONFormats.toBSON(toCache).getOrElse(throw new IllegalArgumentException("Failed to build insert command!")))),
-          modifierForCrudCredentialsBSON(time.getMillis),
-          setOnInsert(BSONDocument(Id -> BSONString(id.id)))
-        ).reduceLeft(_ ++ _)
-
-        def upsert = atomicUpsert(findByIdBSON(id.id), modifiers, AtomicId)
-
-        recoverFromViolation(upsert, upsert)
+        withCurrentTime { time =>
+          def upsert: Future[JSONFindAndModifyCommand.FindAndModifyResult] = findAndUpdate(
+            Json.obj(Id -> id.id),
+            Json.obj(
+              "$set" -> Json.obj(s"${Cache.DATA_ATTRIBUTE_NAME}.$key" -> toCache, "modifiedDetails.lastUpdated" -> time),
+              "$setOnInsert" -> Json.obj("modifiedDetails.createdAt" -> time, Id -> id.id, AtomicId -> BSONObjectID.generate())
+            ),
+            upsert = true,
+            fetchNewObject = true
+          ).recoverWith {
+            case e: DefaultJSONCommandError if e.code.contains(11000) => upsert
+          }
+
+          def handleOutcome(outcome: JSONFindAndModifyCommand.FindAndModifyResult): Future[DatabaseUpdate[Cache]] = {
+            (outcome.lastError, outcome.result[Cache]) match {
+              case (Some(error), Some(value)) =>
+                println(value)
+                if (error.n > 0) {
+                  if (error.updatedExisting) {
+                    Future.successful(DatabaseUpdate(null, Updated(null, value)))
```

Comment:
Can we do something similar to what is done in simple-reactivemongo to create a `DatabaseUpdate` https://github.com/hmrc/simple-reactivemongo/blob/master/src/main/scala/uk/gov/hmrc/mongo/AtomicUpdate.scala#L128-L130 using the `le` function https://github.com/hmrc/simple-reactivemongo/blob/master/src/main/scala/uk/gov/hmrc/mongo/AtomicUpdate.scala#L105.  This would avoid the unsafe nulls.

Reply:
In definition of `Updated` first argument is `previouseValue` - it would break expectation created by name.

I think abstraction here is miss used and not longer good fits in this use case. I will made clean up to preserver compatibility.